### DISCUSSION TITLE
How can one match an array with quotes using Varargs?

### DISCUSSION THREAD

**Discussion by hmf** [scala/scala3#20550] [Metaprogramming] [answered]
I posted this question in the Discord meta-programming channel, and have had no luck. So I am trying here. Assume I have the following code:

```scala
    val s1 = ConstArray.testUnapply(Array(1, 2, 3))
```
I would like to match that array in the `testUnapply(Array(1, 2, 3))` macro. The following match using quotes work: 

```scala
      (x: Expr[Array[T]]) match
        case '{ scala.Array.apply($a:Int, $b:Int, $c:Int): scala.Array[Int]}  => 
          println(s"(a, b, c) = ($a, $b, $c)") 
          ???
```
However, I need to deal with a variable number of arguments. So I tried many variations that include: 

```scala
        case '{ scala.Array.apply(${Varargs(Exprs(elems))}: Seq[Int])}  =>  
=>  
          Some(Array(elems*).asInstanceOf[Array[T]])
          ???
```
None of these work. I have also noticed that the following fails:

```scala
        case '{ scala.Array.apply[Int]($a:Int, $b:Int, $c:Int): scala.Array[Int]}  => 
          println(s"(a, b, c) = ($a, $b, $c)") 
          ???
```
Why does this fail? Do I have to add something related to to `ClassTag`? The following also fails:

```scala
      val fct = summon[FromExpr[ClassTag[Int]]]
      (x: Expr[Array[T]]) match
        case '{ scala.Array.apply( ${varArgs} : Expr[Seq[Int]])(using $fct) }  => 
          ???

```
Any help appreciated. 

TIA

**Comment by som-snytt** [comment]
I started an answer on the forum https://users.scala-lang.org/t/scala-3-macros-quotes-how-to-match-scala-array-1-2-3/10046/2?u=som-snytt. I'll follow up here with more code. Arguably code looks better on github.

**Comment by som-snytt** [comment, marked as answer]
The answer was that since the signatures for `Array.apply` are specialized for primitives, it's necessary to match on those signatures.

```
def _arraylen(xs: Expr[Array[?]])(using Quotes): Expr[Int] =
  xs match
  //case '{ Array(${_}: A, (${Varargs(elems)}: Seq[A])*) } => Expr(elems.length + 1)
  case '{ Array(${_}: Int, (${Varargs(elems)}: Seq[Int])*) } => Expr(elems.length + 1)
  case '{ Array[a]((${Varargs(elems)}: Seq[a])*)($_) } => Expr(elems.length)
  case _ => Expr(-1)
```
or matching on the tree shape
```
def _arraylen(xs: Expr[Array[?]])(using Quotes): Expr[Int] =
  import quotes.reflect.*
  xs match
  case '{ Array[a]((${Varargs(elems)}: Seq[a])*)($_) } => Expr(elems.length)
  case _ =>
    xs.asTerm match
    case Inlined(_, _, Apply(_, List(elem, Typed(Repeated(elems, _), _)))) => Expr(elems.length + 1)
    case _ => Expr(-1)
```
